### PR TITLE
CV + Sonar Full Integration

### LINKED ISSUE #400: Sonar Fixes
# Background

The sonar uses the blue robotics ping-360 library to interact with the physical sonar device, however, the sonar code which we use to connect to the sonar has a few issues. The original system which we developed to communicate with the sonar through ros actions was too slow for the quick sonar pings which we need for sonar-CV integration. Now we are switching to the usual rosnode system but there are lots of artifacts currently within the code base. 

The current sonar code is functional but it needs changes before it is ready to be marked as its own package.

# Purpose

These changes will make the sonar system more robust as currently there are a few problems which will hinder the modularity of the system. In the sonar setup, the sonar is currently found through a hardcoded port. This is a problem as the same code will not work on multiple robots or within a different system. The 'SERIAL_PORT_NAME' should be changed to be found through an automated system similar to 'imu.py'. 

The sonar does not run off of a range, rather it has a 'duration', 'number of samples', and 'period' which it uses to get sonar data over a range of points. The duration is the length of one ping and the period is how often the sonar pings. These two together control the effective range of the sonar. The number of samples is the number of pings used throughout the range determined by the other two parameters. Lowering this will make the sonar faster but more granular. The current way these parameters are calculated is through a linear function, however, they should be changed to match the original code in the pingviewer application made by bluerobotics. 

The data we get from the sonar is stored in an array where the columns are the number of samples and the rows are the angles. Visualizing with numpy, the sonar image that is formed is a polar image in a rectangular format. This is confusing but it works for most purposes. For better CV image processing of the gate, however, it should be changed to polar format.

The first .75 meters of the sonar scan are also noise so they should be dynamically filtered out.

# Tasks
- [ ] Find 'SERIAL_PORT_NAME' dynamically
- [ ] Find a better system to determine the period and duration
- [ ] Create a polar image processing function in 'sonar_image_processing'
- [ ] Create a dynamically filtering method so the first few points of the sonar are filtered out
- [ ] Prune extraneous code from the sonar such as the rosaction code (once the node works)
- [ ] Make a new custom message for the sonar to communicate to CV and task planning over. 
- [ ] Need to test the 'to_robot_position' function to make sure that the position taken from the polar image is the same as the physical location of the object.
- [ ] Update the 'get_xy_of_object_in_sweep' method to be more robust for finding a small cluster of points (right now it takes the average which is prone to extraneous values).
- [ ] The sonar is currently started called from motion.launch. If the thrusters aren't turned on then the sonar isn't either as it is connected to the thrusters power. Also - there are issues if the thrusters are not turned on before the docker container is made! The docker needs to be able to see the sonar in the port for it to be forwarded to the container. Eventually, the sonar will be connected to the main powerbus without the need to turn on the thruster but for now the code should not error out if the sonar is not available.
  - [ ] In the 'sonar_publishing' file, it should throw an error but not stop the program from running. 
  - [ ] In the same file, whenever the sonar is called and disabled it should return an error message.
  - [ ] All other files which call the sonar should be able to interperet this error
- [ ] Remove the 'test_scan_and_finding_gate_posts' method from the sonar class (move it somewhere else)
- [ ] Remove the 'test_gate_from_npy_file' and 'test_buoy_from_npy_file' methods as well.
- [ ] In sonar image processing, remake the gate finding method to use a more robust system for finding the gate. Currently it detects all "circular" objects by perimeter and the finds the one which is the most concave. This should be switched to a different system which can accurately find the gate.
- [ ] The node system needs to be setup for task-planning to interact with the gate finding method.
- [ ] If task-planning is still in the gate stage, the CV method should not be able to be called until the gate is found as that could cause some errors finding the gate.
- [ ] in 'sonar_image_processing' remove the find_buoy' method (or change it)
- [ ] Add buoy hit detection
- [ ] Fix FOV camera buoy correlation

### PR DESCRIPTION
**Pull request by ricohasgithub**
Full CV and Sonar integration. Resolves #400 and #410, and code has been tested in two pool tests to ensure correctness; contains the entirety of the sonar branch which had not yet been merged into master so far.

### CODE REVIEW

File: onboard/catkin_ws/src/cv/scripts/depthai_sonar_spatial_detection.py
Comment (Brokemia) [inline onboard/catkin_ws/src/cv/scripts/depthai_sonar_spatial_detection.py:68]: Is this stuff copy and pasted from the non-sonar DepthAI code? If so, could we make this extend the other class, or at least extract the common code somewhere else? 
I was told to tag @ricohasgithub

File: onboard/catkin_ws/src/cv/scripts/depthai_sonar_spatial_detection.py
Comment (Brokemia) [inline onboard/catkin_ws/src/cv/scripts/depthai_sonar_spatial_detection.py:305]: Is there a better way to indicate errors than 0,0

File: onboard/catkin_ws/src/cv/scripts/depthai_sonar_spatial_detection.py
Comment (Brokemia) [inline onboard/catkin_ws/src/cv/scripts/depthai_sonar_spatial_detection.py:311]: I feel like a Point is better if we aren't using the orientation to store any information

File: onboard/catkin_ws/src/cv/scripts/depthai_sonar_spatial_detection.py
Comment (Brokemia) [inline onboard/catkin_ws/src/cv/scripts/depthai_sonar_spatial_detection.py:323]: Even if sonar gives an error by returning (0, 0), this might still think it's using sonar

File: onboard/catkin_ws/src/cv/scripts/depthai_sonar_spatial_detection.py
Comment (Brokemia) [inline onboard/catkin_ws/src/cv/scripts/depthai_sonar_spatial_detection.py:368]: This is already being done in the above method

File: onboard/catkin_ws/src/cv/scripts/depthai_sonar_spatial_detection.py
Comment (Brokemia) [inline onboard/catkin_ws/src/cv/scripts/depthai_sonar_spatial_detection.py:400]: Put 1.5 or 1.75 as a constant somewhere

File: onboard/catkin_ws/src/sonar/README.md
Comment (Brokemia) [inline onboard/catkin_ws/src/sonar/README.md:1]: image

File: onboard/catkin_ws/src/sonar/scripts/sonar.py
Comment (Brokemia) [inline onboard/catkin_ws/src/sonar/scripts/sonar.py:19]: Is this something that needs to be configurable?

File: onboard/catkin_ws/src/sonar/scripts/sonar.py
Comment (Brokemia) [inline onboard/catkin_ws/src/sonar/scripts/sonar.py:22]: Also configurable? Maybe with the same method as finding the DVL and IMU?

File: onboard/catkin_ws/src/sonar/scripts/sonar.py
Comment (Brokemia) [inline onboard/catkin_ws/src/sonar/scripts/sonar.py:104]: Doesn't default to anything but says it does

File: onboard/catkin_ws/src/sonar/scripts/sonar.py
Comment (Brokemia) [inline onboard/catkin_ws/src/sonar/scripts/sonar.py:170]: \_description\_

File: onboard/catkin_ws/src/sonar/scripts/sonar.py
Comment (Brokemia) [inline onboard/catkin_ws/src/sonar/scripts/sonar.py:231]: Could tests get their own file?

File: onboard/catkin_ws/src/sonar/scripts/sonar.py
Comment (Brokemia) [inline onboard/catkin_ws/src/sonar/scripts/sonar.py:285]: Might be good README material

File: onboard/catkin_ws/src/sonar/scripts/sonar_image_processing.py
Comment (Brokemia) [inline onboard/catkin_ws/src/sonar/scripts/sonar_image_processing.py:51]: Why is this all caps, it looks like a constant

File: onboard/catkin_ws/src/sonar/scripts/sonar_publisher.py
Comment (Brokemia) [inline onboard/catkin_ws/src/sonar/scripts/sonar_publisher.py:9]: Should this be configurable?

**Review by Brokemia — CHANGES_REQUESTED** [review]
Don't forget to remove the random images

File: onboard/catkin_ws/src/cv/scripts/depthai_sonar_spatial_detection.py
Comment (ricohasgithub) [reply onboard/catkin_ws/src/cv/scripts/depthai_sonar_spatial_detection.py:68]: That's a good point. I think the code is slightly different as we took out some stuff with stereo that we won't need due to the use of sonar to help speed things up. We can definitely consider this in a future update.

File: onboard/catkin_ws/src/cv/scripts/depthai_sonar_spatial_detection.py
Comment (ricohasgithub) [reply onboard/catkin_ws/src/cv/scripts/depthai_sonar_spatial_detection.py:323]: Currently in_sonar_range will only ever be true if sonar is running and in range, so the only case for if sonar_response is (0, 0) is when the sonar has not responded, which in case in_sonar_range will be False.

**Review by Brokemia — APPROVED** [review]
Looks good now

**Comment by vedarshshah** [comment]
Please hold off on merging this until #411 and #402 have been merged in as their changes must be merged into the CV portion of this branch.

**Pull request merged**
